### ISSUE TITLE
Configuration option to specify branch to use besides default

### ISSUE THREAD

**Issue by samjhill** [dabutvin/Imgbot#155]
Thank you very much for creating this bot and releasing it open-source! I really appreciate it.

Is there currently any config setting where we can pass in a branch name other than `master`? We work from a staging branch, which gets merged into master when approved. It'd be awesome if imgbot would make PRs against staging, so we don't have to merge master back into staging just for images.

Thanks!
Sam

**Comment by dabutvin** [comment]
Hi! Thanks for stopping by and opening an issue.

I've wanted to do something like this for a while, but haven't quite landed on the proper design yet.

May I ask, why don't you set the branch you work from as the default branch?

ImgBot is not tied to the branch named 'master' it is tied to the branch designated as 'default'

**Comment by leemengtaiwan** [comment]
Thanks for building the app and I also hope we can specify which branch for the bot to look for image compression as well.

In my use-case, I have a Github page which serve content at the `master` branch, while these content are built from `dev` branch.

So even Imagebot make a nice [PR](https://github.com/leemengtaiwan/leemengtaiwan.github.io/pull/5) for reducing image sizes of my `master` branch, these updates will get reset if I build the content from `dev` branch again 😿

**Comment by Berkmann18** [comment]
@dabutvin The problem with changing the default branch so that ImgBot creates a PR against the correct one (say `dev`) is that ot will switch what some people use as the release/stable branch (usually `master`) which is what is displayed by _default_ on the repos with the staging/work/development branch which could cascade in having other tools linked to the repo as not working on the proper branches.

Also, ImgBot doesn't seem to be able to pick up the commit convention which would ruin the changelog generation and consistency in the commit messages.

**Comment by Berkmann18** [comment]
I've opened #232 regarding the commit convention with some examples.

**Comment by dabutvin** [comment]
So we can get the .imgbotconfig to read the branch name to use besides default with the contents api
for example: https://api.github.com/repos/dabutvin/imgbot/contents/.imgbotconfig

The tricky part will be making this work for private repositories as we don't have an installation token at this point, so we will have to get a token first. Probably want to do that anyways to help avoid rate limiting using the github api so often.

I think it will be best to check the affected files first, before checking the branch name, as that will be much cheaper after this change.

**Comment by dabutvin** [comment]
- [ ] update webhook to respect imgbotconfig for branch changes
- [ ] update compress routine to checkout the correct branch
- [ ] update the pull request routine to compare imgbot branch to the configured branch

**Comment by dabutvin** [comment]
One flaw I can see in this design, will the `.imgbotconfig` live in the default branch? How will we know where to look for this file if it is in a non-default branch? Is that a problem?

**Comment by Berkmann18** [comment]
> One flaw I can see in this design, will the `.imgbotconfig` live in the default branch? How will we know where to look for this file if it is in a non-default branch? Is that a problem?

I guess, the sensible thing to do would be to first look at the default branch for it then if it's not there, either: check other __active__ branches.

Other than that,  think it would be sensical to expect the bot to only look at the default branch and once it knows that the `.imgbotconfig` (or `.imgbotrc` or whatever name it can accept) it will parse it and act accordingly.

**Comment by dabutvin** [comment]
I wonder if it would be easier to make this configuration something you would do at https://imgbot.net/app

Find your repo and type in the branch you want to check as a setting and we can store it outside the source code with your installation metadata.

Something about putting a `.imgbotconfig` file in your default branch when you want to point to another branch feels weird to me.

**Comment by danner26** [comment]
I definitely think this would be a good thing to have at the https://imgbot.net/app website. 

This is a setting that for me, on some repos I want it only touching master, but there are exceptions like my latest project which master is not the branch I want imagebot to touch as it doesn't get pushed to much, and it doesn't contain the up to date photos/folder structure because of that.. 

I think that the only way to use the .imgbotconfig file would be to require users to make the file in the master branch, which may work for some but not for everyone.

For instance, one of our policies is that nothing gets committed directly to master, and this would break our ruleset. 

So, this is why I believe that having this as a configurable option on the control website will be a much better idea. (in reference to #354 )

**Comment by jasikpark** [comment]
I would agree that specifying the branches it operates on per-repository should be an app-setting rather than a configuration file. Unless if the bot would look in **every** non-master branch for the configuration file every time it runs.

**Comment by RenanMsV** [comment]
So then you should move all configs into the app. Its one or another. Not both.

**Comment by dabutvin** [comment]
I don't want to lose this one.
To move this forward, I think we can leave the .imgbotconfig settings out of it for now and offer a branch setting right from the website where you log in.

I agree with @RenanMsV that having two places for config is not ideal, but I'm not sure how much I like moving everything to the website right now.

**Comment by Remzi1993** [comment]
Hi @dabutvin What's the status of this issue? I would like to know if this is going to be implemented in the future? I'm just curious. 

I ran into a problem with a forked repo and doing a PR to the upstream repo, because I also use a [Pull App](https://github.com/apps/pull) to keep my default branch up to date with the upstream repo. 

To work around this I needed to create a separate branch and set this temporary the default branch. To select or create a branch is a much desired functionality especially for forked repo's working with the fork an pull model.

Kind regards,
@Remzi1993

**Comment by dabutvin** [comment]
Hi @Remzi1993 I really want to do this one, but I want to make sure it gets done in the right way.

We don't need any more justification as this is the most common issue being run into with imgbot these days. Now we just need to do the work.

The major part that needs work is adding a settings ui to the website frontend at https://github.com/dabutvin/ImgBot/blob/master/Web/src/app/vue/components/Repository.vue accessed from https://imgbot.net/app

**Comment by Remzi1993** [comment]
Hi @dabutvin thanks for your reply. I wish I could help you, but I just a junior PHP dev.
I wish you good luck 👍  Thank you for this awesome bot 🎉

**Comment by dabutvin** [comment]
the work for this is all done now!
see PR #497 
Just need to do some testing before we can merge and release

**Comment by Remzi1993** [comment]
@dabutvin Thank you very much!

**Comment by dabutvin** [comment]
This is now live in production. You can update your repo's setting by logging into https://imgbot.net/app and hitting the gear icon.

Please reach out if you run into any problems

**Comment by dabutvin** [comment]
Also, don't forget, if you have an existing PR into your default branch and you are changing over to use an override, don't forget to blow away the existing imgbot branch and pr so the not has somewhere to work

**Comment by Maruda1986** [comment]
> Thank you very much for creating this bot and releasing it open-source! I really appreciate it.
> 
> Is there currently any config setting where we can pass in a branch name other than `master`? We work from a staging branch, which gets merged into master when approved. It'd be awesome if imgbot would make PRs against staging, so we don't have to merge master back into staging just for images.
> 
> Thanks!
> Sam

**Comment by jasikpark** [comment]
Thanks a lot!!
